File: OscilloscopeGraphics/Clipper.cpp
#include "Clipper.h"

using namespace osc;

ClipPolygon::~ClipPolygon() { delete normals; }

void ClipPolygon::calculateNormals() {
  for (int i = 0; i < vertexCount; i++) {
    Point dir;
    (vertices[(i + 1) % vertexCount]).sub(vertices[i], dir);
    normals[i].x = -dir.y;
    normals[i].y = dir.x;
    normals[i].normalize();
  }
};

/*
 * Clip line against a convex polygon
 *
 * Cyrus-Beck line clipping algorithm implementation:
 * https://en.wikipedia.org/wiki/Cyrus%E2%80%93Beck_algorithm
 */
bool Clipper::clipLine(Line& line) {
  float tE = 0, tL = 1;
  Point P1_P0, PEi;
  line.b.sub(line.a, P1_P0);

  for (int i = 0; i < viewport->vertexCount; i++) {
    line.a.sub(viewport->vertices[i], PEi);
    float num = viewport->normals[i].dot(PEi);
    float den = viewport->normals[i].dot(P1_P0);

    if (den != 0) {
      float t = num / -den;

      if (den < 0) {
        tE = maxValue(tE, t);  // Entering
      } else {
        tL = minValue(tL, t);  // Exiting
      }
    } else if (num > 0) {  // Parallel and outside
      return false;
    }
  }

  if (tE > tL) {  // Outside
    return false;
  }

  float x0 = line.a.x;
  float y0 = line.a.y;

  line.a.x = x0 + P1_P0.x * tE;
  line.a.y = y0 + P1_P0.y * tE;
  line.b.x = x0 + P1_P0.x * tL;
  line.b.y = y0 + P1_P0.y * tL;
  return true;
}

float Clipper::minValue(float a, float b) { return a < b ? a : b; }

float Clipper::maxValue(float a, float b) { return a > b ? a : b; }

File: osc/clipper.py
from dataclasses import dataclass, astuple
from math import sqrt
from typing import NamedTuple, Optional


class Point(NamedTuple):
    x: float
    y: float


class Line(NamedTuple):
    a: Point
    b: Point
    z: float


def sub(a: Point, b: Point) -> Point:
    return Point(a.x - b.x, a.y - b.y)


def dot(a: Point, b: Point) -> float:
    return a.x * b.x + a.y * b.y


def normalize(p: Point) -> Point:
    length = sqrt((p.x ** 2) + (p.y ** 2))
    return Point(p.x / length, p.y / length)


def calculate_normal(a: Point, b: Point) -> Point:
    x, y = sub(a, b)
    return normalize(Point(-y, x))


class Polygon:
    def __init__(self, points=[Point(-1.0, 0.75), Point(1.0, 0.75), Point(1.0, -0.75), Point(-1.0, -0.75)]):
        self.points = points
        self.normals = self._calculate_normals(self.points)

    def _calculate_normals(self, points):
        normals = []
        for i in range(len(points)):
            x, y = (sub(points[(i + 1) % len(points)], points[i]))
            normals.append(normalize(Point(-y, x)))

        return normals


#
# Clip line against a convex polygon
#
# Cyrus-Beck line clipping algorithm implementation:
# https://en.wikipedia.org/wiki/Cyrus%E2%80%93Beck_algorithm
#
def clip_line(line: Line, polygon: Polygon) -> Optional[Line]:
    a, b, z = line
    [te, tl] = [0.0, 1.0]

    p1p0 = sub(b, a)

    for i in range(len(polygon.points)):
        pei = sub(a, polygon.points[i])
        num = dot(polygon.normals[i], pei)
        den = dot(polygon.normals[i], p1p0)

        if den != 0:
            t = num / -den
            if (den < 0):
                te = max(te, t)  # Entering
            else:
                tl = min(tl, t)  # Exiting
        elif num > 0:  # Parallel and outside
            return None

    if te > tl:  # Outside
        return None

    return Line(
        Point(a.x + p1p0.x * te, a.y + p1p0.y * te),
        Point(a.x + p1p0.x * tl, a.y + p1p0.y * tl),
        z
    )
